Tools/clinic/clinic.py

```diff
@@ -2577,15 +2588,15 @@ class CConverter(metaclass=CConverterAutoRegister):
 
     # If not None, default must be isinstance() of this type.
     # (You can also specify a tuple of types.)
-    default_type = None
+    default_type: Type[Any] | tuple[Type[Any], ...] | None = None
```

Comment:
Would prefer using `builtins.type` here

Reply:
Oh, do you mean the fully qualified form? I.e.

```py
import builtins
default_type: builtins.type[Any] | tuple[builtins.type[Any], ...] | None = None
```

Reply:
Relatedly do you (or @erlend-aasland) know what the Python version support for clinic is supposed to be? I think we have to support running it with at least 3.11 so you can use it if you messed up your current build. But maybe we need to go back to 3.7 or 3.8, in which case a few of the annotations in this PR would fail.

Reply:
> Relatedly do you (or @erlend-aasland) know what the Python version support for clinic is supposed to be? I think we have to support running it with at least 3.11 so you can use it if you messed up your current build. But maybe we need to go back to 3.7 or 3.8, in which case a few of the annotations in this PR would fail.

Yes, I wondered about that. In https://github.com/python/cpython/pull/104322, I noticed that it looked like some parts of `clinic.py` were taking pains so that they could be run on Python 3.7. Now that that PR is merged, it definitely can't be run on Python 3.7 anymore :)

The abstract at the top of Argument Clinic basically says "Feel free to use Argument Clinic outside of CPython, but do so 100% at your own risk, there are no compatibility guarantees: https://docs.python.org/3/howto/clinic.html. So I think that means that we _should_ feel free to abandon support for Python <3.11 if we want to. But I'm hardly an expert on Argument Clinic!!

Reply:
In my `./configure` script `python3.10` get selected by default.

Reply:
The only typing feature from 3.11 that we might want to use at some point in `clinic.py` would be `typing.Self`. But we haven't got any uses of that right now in `clinic.py`.

Reply:
I'd still like to hear Erlend's opinion on version support. `configure` theoretically could pick a Python as low as 3.6 (check for `PYTHON_FOR_REGEN` in configure.ac). I don't think we need to support 3.6, but 3.10 may be too aggressive.

Reply:
If necessary, we could probably use `from __future__ import annotations` in order to use modern syntax while maintaining compatibility with Python 3.8+.

Reply:
I'm not sure Larry would approve if we did that to his tool :D

Reply:
> In my `./configure` script `python3.10` get selected by default.

Do you use a configure cache? It should select `python3.11` if you've got that in your `PATH`.

> I'd still like to hear Erlend's opinion on version support. `configure` theoretically could pick a Python as low as 3.6 (check for `PYTHON_FOR_REGEN` in configure.ac). I don't think we need to support 3.6, but 3.10 may be too aggressive.

Yes, `PYTHON_FOR_REGEN` (which is used for stuff like `make clinic`) can theoretically be `python3.6`[^1]. As Alex correctly points out, clinic.py already requires 3.7. Since 3.6 is now EOL (IIRC), perhaps we can raise `PYTHON_FOR_REGEN` to 3.7, and try to keep clinic.py 3.7 compatible. Let's do some research about which Python 3 version is shipped with the popular distros nowadays (distrowatch.org or a similar database).

[^1]: but it also checks for plain `python3` and even `python`, so I guess it could be anything

Reply:
> As Alex correctly points out, clinic.py already requires 3.7. Since 3.6 is now EOL (IIRC), perhaps we can raise `PYTHON_FOR_REGEN` to 3.7, and try to keep clinic.py 3.7 compatible.

Correction: following #104322, clinic now requires 3.8+.

> Since 3.6 is now EOL (IIRC)

Yes, 3.6 is EOL, and 3.7 is about to become EOL

Reply:
Exciting news: running mypy on clinic.py has already discovered a real, genuine bug! If anybody had tried running clinic.py with 3.8, they would have been hit by multiple `NameError` exceptions. All these `NoneType` references are undefined on Python 3.8:

https://github.com/python/cpython/blob/c527eb1c2a473df01c19195b378f780e9180fd1c/Tools/clinic/clinic.py#L3276

https://github.com/python/cpython/blob/c527eb1c2a473df01c19195b378f780e9180fd1c/Tools/clinic/clinic.py#L3303-L3306

https://github.com/python/cpython/blob/c527eb1c2a473df01c19195b378f780e9180fd1c/Tools/clinic/clinic.py#L3439

https://github.com/python/cpython/blob/c527eb1c2a473df01c19195b378f780e9180fd1c/Tools/clinic/clinic.py#L3459

https://github.com/python/cpython/blob/c527eb1c2a473df01c19195b378f780e9180fd1c/Tools/clinic/clinic.py#L3552-L3553

https://github.com/python/cpython/blob/c527eb1c2a473df01c19195b378f780e9180fd1c/Tools/clinic/clinic.py#L3593

https://github.com/python/cpython/blob/c527eb1c2a473df01c19195b378f780e9180fd1c/Tools/clinic/clinic.py#L3613-L3629

https://github.com/python/cpython/blob/c527eb1c2a473df01c19195b378f780e9180fd1c/Tools/clinic/clinic.py#L3654

https://github.com/python/cpython/blob/c527eb1c2a473df01c19195b378f780e9180fd1c/Tools/clinic/clinic.py#L3673

https://github.com/python/cpython/blob/c527eb1c2a473df01c19195b378f780e9180fd1c/Tools/clinic/clinic.py#L3687

On Python 3.10+, `NoneType` is defined due to the `from types import *` line at the top of the file. But `types.NoneType` is new in Python 3.10 -- it doesn't exist on earlier versions of Python!

Reply:
So, I guess 3.8 is just not supported. Maybe 3.10+?

Reply:
> So, I guess 3.8 is just not supported. Maybe 3.10+?

It's trivial to support 3.8, though. All we need to do is just add `NoneType = type(None)` at the top.

Reply:
> > So, I guess 3.8 is just not supported. Maybe 3.10+?
> 
> It's trivial to support 3.8, though. All we need to do is just add `NoneType = type(None)` at the top.

Hmm, it was slightly fiddlier than I thought. 9fcd87fd09247c5926cfcde18cabc1c35dc90356 is the changes required to make `clinic.py` compatible with Python 3.8 again.

Reply:
FYI @erlend-aasland I haven't tried bumping the `PYTHON_FOR_REGEN` variable (and don't know how to), but happy to do that in this PR if you point me to which code needs to be changed! Alternatively, I'm also happy to leave that to a followup PR :)

Reply:
That should be done in a separate PR, but f7241aaf6f22fdd4710f8d3ef1d594f9725b3b90 should have the right files edited.

Reply:
I took a quick look at distrowatch, and I think requiring 3.8 for `PYTHON_FOR_REGEN` is fine[^1]. Let's bump the build system in a separate PR, as Jelle says.

You can use `PYTHON_FOR_REGEN=python3.8 ./configure && make` to test this locally.

[^1]: others might be more conservative than me, of course

Reply:
I might be wrong, but judging from the tiny bits of typing knowledge I've acquired, I think `Tools/cases_generator/generate_cases.py` (which is run by `PYTHON_FOR_REGEN`) is not compatible with typing as of 3.8.

Reply:
Sorry for spamming, but AFAICS, the newest Python we can do a `make regen-all` with, is Python 3.10. So, either we'll change `configure` to require that, or we adjust the tools. Let's move this discussion to Discourse.

Reply:
> Oh, it also uses pattern matching. It is definitely impossible to `make regen-all` with `PYTHON_FOR_REGEN=python3.8`.

Huh. Shall I revert 9fcd87f, in that case? I don't much like the `from _typeshed import NoneType` hack, at all.

Reply:
> the newest Python we can do a `make regen-all` with, is Python 3.10.

You mean the _oldest_, right? :)

Reply:
Yes, yes, yes 😆

Reply:
FTR, I was able to do `make PYTHON_FOR_REGEN=python3.8 regen-all` both in 3.11 and 3.10 builds.

Reply:
> [gh-104487](https://github.com/python/cpython/issues/104487) settles on requiring Python 3.10 for regen stuff, so... use all the new stuff :)

Fab -- done!